---
  html_document:
    toc: true
    highlight: zenburn
    theme: united
---

```{r setup, echo=FALSE}
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png",
               cache=TRUE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE,
               message=FALSE, prompt=TRUE, comment='', fig.cap='')
```

# Overview
This was a pilot project to see if it was possible to pick out a subpopulation
of single HSC cells. About 50 MPP and 50 HSC single cells were prepped using
the SMART-seq protocol and the reads were sequenced to a depth of a couple million
per cell. This is the second shot at this experiment, previously we had done some
work with SCRB-seq sequencing which had more cells but at a lower depth and
a lower library complexity.

```{r qc-setup, echo=FALSE}
library(ggplot2)
library(reshape)
library(gplots)
library(edgeR)
library(CHBUtils)
library(pheatmap)
project_summary = "/Users/rory/cache/agnieszka-singlecell/results/project-summary.csv"
counts_file = "/Users/rory/cache/agnieszka-singlecell/results/combined.counts"
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442",
"#0072B2", "#D55E00", "#CC79A7")
summarydata = data.frame(read.table(project_summary, header=TRUE, sep=","), row.names="Name", check.rows=FALSE)
summarydata$Name = rownames(summarydata)
summarydata = summarydata[order(summarydata$Name),]
counts = read.table(counts_file, header=TRUE, row.names="id", check.names=FALSE)
counts = counts[, order(colnames(counts))]
colnames(counts) = gsub(".counts", "", colnames(counts))
# this is a list of all non user-supplied metadata columns that could appear
known_columns = c("Name", "X.GC", "Exonic.Rate", "Sequences.flagged.as.poor.quality",
    "rRNA.rate", "Fragment.Length.Mean", "Intronic.Rate", "Intergenic.Rate",
    "Quality.format", "Duplication.Rate.of.Mapped", "Mapped",
    "rRNA", "Sequence.length", "Transcripts.Detected", "Mean.Per.Base.Cov.",
    "Genes.Detected", "Unique.Starts.Per.Read", "unique_starts_per_read",
    "complexity", "X5.3.bias")
```

```{r heatmap-function, echo=FALSE}
get_heatmap_fn = function(summarydata) {
    # return the pheatmap function with or without metadata
    metadata = summarydata[, !colnames(summarydata) %in% known_columns, drop=FALSE]
    if(ncol(metadata) == 0) {
       return(pheatmap)
    }
    else {
    # rownames(metadata) = summarydata$Name
    heatmap_fn = function(data, ...) {
        pheatmap(data, annotation=metadata, ...)
    }
    return(heatmap_fn)
}}
heatmap_fn = get_heatmap_fn(summarydata)
```

# Quality control metrics

## Mapped reads
```{r mapped-plot, echo=FALSE}
ggplot(summarydata, aes(x=Name, y=Mapped)) +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    geom_bar(stat="identity") +
    ylab("mapped reads") + xlab("")
```

The number of mapped reads varies among the samples but is within the range we
were expecting, with an average of `r mean(summarydata$Mapped)` reads per
sample. It is natural to have some variation here.

## Genomic mapping rate
```{r mapping-rate-plot}
ggplot(summarydata, aes(x=Name, y=Mapping.Rate)) +
    geom_bar(stat="identity") +
    ylab("mapping rate") + xlab("") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90))
```

The average genomic mapping rate is `r mean(summarydata$Mapping.Rate)`, which
looks great. There are a couple cells that have very low mapping rates, likely
those are cells that failed somehow, either there was an empty well or there was
a problem with the library prep for that cell or something similar. This is a
pretty common occurance.

## Number of genes detected
```{r genes-detected-plot}
dd = data.frame(Name=names(counts), Genes.Detected = colSums(counts > 0))
ggplot(dd, aes(x=Name, y=Genes.Detected)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("genes detected") + xlab("")
```

We can detect on average of `r mean(dd$Genes.Detected)` genes detected per
cell, using a permissive definition of detected meaning there is at least one
read in the gene.

## Gene detection saturation
```{r saturation-plot}
dd = data.frame(Mapped=summarydata$Mapped, Genes.Detected = colSums(counts > 0))
ggplot(dd, aes(x=Mapped, y=Genes.Detected)) +
    geom_point() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("genes detected") + xlab("reads mapped")
```

The idea behind this plot is we can look at it and try to figure out if sequencing
deeper would let us detect more genes. Here we can see that sequencing less
than a million reads has a big dropoff for the number of genes detected. It
looks like there is a pretty linear increase in genes detected other than that.
However, sequencing six million reads instead of two million reads and
sequencing around two million reads per sample is around the sweet spot
for maximizing the number of genes we can detect.

## Exonic mapping rate
```{r exonic-mapping-plot}
ggplot(summarydata, aes(x=Name, y=Exonic.Rate)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("exonic mapping rate") + xlab("")
```

Less good is the exonic mapping rate, we have a lower number of reads
mapping to exons. This indicates that although a high amount of the reads
align to the genome, there is likely some genomic contamination, so the
reads are not aligning where there are genes.

## Estimated fragment length of paired-end reads
```{r fragment-length-plot}
ggplot(summarydata, aes(x=Name, y=Fragment.Length.Mean)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("fragment length") + xlab("")
```

The estimated fragment length is very small, this is an estimate of the size of
the piece of RNA that was between the adapters. A tiny fragment size means the
RNA was likely sheared too much in the library prepartion process, leading to
many small fragments.

## Boxplot of log10 counts per gene
```{r boxplot-raw}
melted = melt(counts)
colnames(melted) = c("sample", "count")
melted$sample = factor(melted$sample)
melted = melted[order(melted$sample),]
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

The overall distribution of counts per cell looks pretty nice, especially for
a single cell dataset.

## Boxplot of log10 TMM-normalized counts per gene
```{r boxplot-normalized}
y = DGEList(counts=counts)
y = calcNormFactors(y)
normalized_counts = cpm(y, normalized.lib.sizes=TRUE)
melted = melt(normalized_counts)
colnames(melted) = c("gene", "sample", "count")
melted$sample = factor(melted$sample)
melted = melted[order(melted$sample),]
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

Despite it being a single-cell dataset, TMM normalizing the reads does a nice job
normalizing the cells so they have similar distributions.


## Filter out unexpressed genes and samples with low total counts
```{r filter-low-counts}
condition = "cell"
keep_cells = colSums(counts) > 1e4
counts = counts[rowSums(counts) > 0,]
counts = counts[, keep_cells]
summarydata = summarydata[colnames(counts),]
groups = summarydata[,condition]
```

## Totally naive HSC-MPP DE analysis
Looking at just the differences between the HSC and MPP cell types, we
can find some global differences.
```{r scde, eval=!file.exists("scde.RData")}
library(scde)
n.cores = 1
o.ifm = scde.error.models(counts=counts, groups=groups, n.cores=n.cores,
                          threshold.segmentation=T, save.crossfit.plots=F,
                          save.model.plots=F, verbose=1)
valid.cells = o.ifm$corr.a > 0
o.ifm = o.ifm[valid.cells,]
groups = groups[valid.cells]
counts = counts[valid.cells,]
o.prior = scde.expression.prior(models=o.ifm, counts=counts, length.out=400,
                                show.plot=F)
ediff <- scde.expression.difference(o.ifm, counts, o.prior, groups=groups,
                                    n.randomizations=100,
                                    n.cores=n.cores, verbose=1)
ediff$pvalue = pnorm(-(abs(ediff$Z))) * 2
ediff$padj = p.adjust(ediff$pvalue)
ediff = ediff[order(ediff$pvalue),]
save(o.ifm, groups, counts, o.prior, ediff, file="scde.RData")
```

```{r add-symbol}
load("scde.RData")
library(biomaRt)
mouse = useMart("ensembl", dataset="mmusculus_gene_ensembl")
conversions = getBM(attributes=c("ensembl_gene_id", "mgi_symbol"), mart=mouse)
matches = match(rownames(ediff), conversions$ensembl_gene_id)
ediff$symbol = conversions$mgi_symbol[matches]
write.table(ediff, file="singlecell_de.tsv", sep="\t", quote=FALSE,
            row.names=TRUE, col.names=TRUE)
```

## Marker exploration
Here we look at some of the markers that were used to sort to see if
we can see the RNA-seq data matches the sorting markers.

```{r seurat}
library(Seurat)
scounts = counts
scounts$id = rownames(scounts)
scounts = merge(scounts, conversions, by.x="id", by.y="ensembl_gene_id")
scounts = scounts[!duplicated(scounts$mgi_symbol),]
rownames(scounts) = scounts$mgi_symbol
scounts$mgi_symbol = NULL
scounts$ensmbl_gene_id = NULL
scounts$id = NULL
scounts = scounts[rowSums(scounts > 0) > 5,]
keep = colSums(scounts) > 10000

nbd.data = data.frame(log(scounts + 1))
cnames = paste(colnames(nbd.data), groups, sep="_")
colnames(nbd.data) = cnames
nbt = new("seurat", raw.data=nbd.data)
nbt = Seurat:::setup(nbt, project="NBT", min.cells = 3, min.genes = 1000, is.expr=1,
           names.delim="_", names.field=2)
```

Here we look at Weissman's markers for LT-HSC, ST-HSC, Early-MPP and Late-MPP.

LT-HSC : CD34−, CD38−, SCA-1+, Thy1.1+/lo, C-kit+, lin−, CD135−, Slamf1/CD150+

ST-HSC : CD34+, CD38+, SCA-1+, Thy1.1+/lo, C-kit+, lin−, CD135−, Slamf1/CD150+, Mac-1 (CD11b)lo

Early MPP : CD34+, SCA-1+, Thy1.1−, C-kit+, lin−, CD135+, Slamf1/CD150−, Mac-1 (CD11b)lo, CD4lo

Late MPP : CD34+, SCA-1+, Thy1.1−, C-kit+, lin−, CD135high, Slamf1/CD150−, Mac-1 (CD11b)lo, CD4lo

Here we look at the Slam markers for mouse, according to wikipedia
which is:

Mouse HSC : CD34lo/−, SCA-1+, Thy1.1+/lo, CD38+, C-kit+, lin−

SCA-1 is Atxn1.

```{r weissman-markers}
vlnPlot(nbt, c("Cd34", "Cd38", "Atxn1", "Thy1", "Kit", "Flt3", "Slamf1"))
```


```{r slam-markers}
vlnPlot(nbt, c("Cd34", "Slamf1", "Cd244", "Cd48", "Cd38", "Flt3"))
```

They are supposed to be positive for Cd38 but aren't, which is concerning.

```{r fgd5-marker}
vlnPlot(nbt, c("Fgd5", "Runx1t1", "Neo1", "Igf1", "Mpo", "Cd48", "Cd244",
               "Bex6"))

```
Cd4 and Cd8a and Il7r and Ly6g all have symbols just aren't expressed.
Ter119 is Ly76, we don't have that in the annotation.
B220 is an isoform of Cd45 and we are aggregating at the gene level so we
don't have that data here.
Mac1 is Itgam and Itgb2.

The lin- markers seem good.

HSC = Lin-Ckit+Sca1+CD34-CD135-CD150+

MPP-1 = Ckit+Lin-Sca1+CD34+CD135- (85% cells CD150- but did not sort on this)

Sca1 is Ly6a
Cd135 is Flt3
Cd150 is Slamf1

```{r digital-sorting}
vlnPlot(nbt, c("Kit", "Ly6a", "Cd34", "Flt3", "Slamf1", "Cd38"))
anno = data.frame(cell=groups)
rownames(anno) = colnames(scounts)

markers = c("Kit", "Ly6a", "Cd34", "Flt3", "Slamf1", "Cd38")
pheatmap(log(scounts[markers,] + 1), fontsize=6, annotation=anno)

library(tidyr)
df = scounts
df$id = rownames(df)
df = melt(df[markers,]) %>%
     separate("variable", into=c("well", "type"), sep="_")
ggplot(df, aes(well, value, fill=type, color=type)) +
   geom_bar(stat='identity', position='dodge') + facet_wrap(~ id) +
   xlab("") +
   theme_bw() +
   theme(axis.ticks = element_blank(), axis.text.x = element_blank())
```


## Dirty microarray analysis
I can't do a proper microarray analysis without the original CEL files, so here
I just loaded the microarray values in and performed a differential expression
test for each probeset with limma. I didn't normalize or anything so this is a
super dirty list, but it is probably a lot better than just looking at fold changes.

```{r microarrays}
microarray = read.table("/Volumes/Clotho/Users/rory/cache/agnieszka-singlecell/metadata/Rossi_HSC_MPP_Microarray.csv", header=TRUE, sep=",", stringsAsFactors=FALSE)
probe_symbol = microarray[, c(1, 2)]
rownames(microarray) = microarray[, 1]
microarray = microarray[, c(6, 7, 8, 9, 12, 13, 14, 15, 16)]
colnames(microarray) = c("HSC_1", "HSC_2", "HSC_3", "HSC_4", "MPP_1", "MPP_2", "MPP_3", "MPP_4", "MPP_5")
microarraysamples = data.frame(type=c("HSC", "HSC", "HSC", "HSC", "MPP", "MPP", "MPP", "MPP", "MPP"))
rownames(microarraysamples) = colnames(microarray)
microarray = log(microarray)
design = model.matrix(~ type, data=microarraysamples)
fit = lmFit(microarray, design)
fit2 = eBayes(fit)
mresults = topTable(fit2, adjust="BH", p.value=1, n=Inf)
mresults = merge(mresults, probe_symbol, by.x="row.names", by.y="Probeset")
rownames(mresults) = mresults$Row.names
mresults$Row.names = NULL
mresults = mresults[order(mresults$adj.P.Val),]
```

There is a very weak correlation between the log fold change in the gene
array and the log fold change in the singlecell RNA-seq data.

```{r singlecell-vs-microarray}
in_both = intersect(ediff$symbol, mresults$Gene.Symbol)
z = merge(ediff, mresults, by.x="symbol", by.y="Gene.Symbol")

ggplot(z, aes(mle, -logFC)) + geom_point() + xlab("RNA-seq") + ylab("microarray") +
    theme_bw()
```

However, there is a reasonable correlation between the fold change in the
sorting genes between the microarray and RNA-seq data:

```{r rnaseq-vs-microarray-markers}
library(devtools)
source_gist("524eade46135f6348140")
ggplot(subset(z, symbol %in% markers), aes(mle, -logFC, label=mle)) +
geom_point() + xlab("RNA-seq") + ylab("microarray") +
    theme_bw() + stat_smooth_func(geom="text", method="lm", parse=TRUE, hjust=0) +
    geom_smooth(method="lm", se=FALSE)
```

## PCA of single cell data
If there are subpopulations in the single-cell data we should be able to see
them in a PCA plot. Here we tae the most variable genes between the datasets
and perform PCA on them to see if they separate into groups.

```{r identify-variable-genes}
nbt = mean.var.plot(nbt, y.cutoff=2, x.low.cutoff=2, fxn.x=expMean,
                    fxn.y=logVarDivMean)
length(nbt@var.genes)
```
PCA of components 1 and 2 doesn't do a great job of separating out the
HSC an MPP cells. It sort of separates some of the MPP and HSC cells
but there are a set of MPP cells that cluster with the HSC cells. The
HSC cells don't appear to have any well defined subpopulations
of cells.

```{r pca-jackstraw}
nbt = pca(nbt, do.print=FALSE)
pca.plot(nbt, 1, 2)
```

Here are the genes that make up the first and second components:

```{r component-1-genes}
print.pca(nbt, 1)
```

```{r component-2-genes}
print.pca(nbt, 2)
```

It looks like there is maybe some heterogeneity in the MPP cells but
there isn't very much heterogeneity in the HSC cells in terms of there
being a few clear separate groups.

Looking at higher order components also doesn't separate out the HSC
cells into any subpopulations.

```{r pca-3-4}
pca.plot(nbt, 3, 4)
pca.plot(nbt, 5, 6)
pca.plot(nbt, 7, 8)
```

Just to demonstrate what you'd might expect from a clearer set of multiple
cell types, here is an example dataset from Pollen et al where
you can see a better separation of cells by a quick PCA plot:
http://www.ncbi.nlm.nih.gov/pubmed/25086649

```{r pollen-dataset}
library(scde)
library(Seurat)
data(pollen)
pollen = pollen[, colSums(pollen > 0)>2000]
pollen = pollen[rowSums(pollen) > 10,]
pollen = pollen[rowSums(pollen > 0) > 5,]
pollen.data = data.frame(log(pollen + 1))
pollen.data = pollen.data[, order(colnames(pollen.data))]
x = paste("GW16", 1:25, sep="_")
x = c(x, paste("GW21", 26:49, sep="_"))
x = c(x, paste("NPC", 50:64, sep="_"))
colnames(pollen.data) = x
pr = prcomp(pollen.data)
rot = pr$rotation
 ggplot(data.frame(rot), aes(x=PC1, y=PC2, label=rownames(rot))) + geom_text() +
     theme_bw()
```

I'm not sure what else we can do with this dataset if the main goal is to look
at a subpopulations of HSCs, they are very homogeneous and we don't have very
many cells.

We could look at the populations of MPPs, because it looks like there are three
groups. One group off by itself, another group that clusters with the HSC cells
and another group that is close to the HSC cells but does not cluster with them.

```{r mpp-subpopulations}
z = nbt@pca.rot
z = z[grepl("MPP", rownames(z)),]
mpp1 = rownames(z[z$PC1 < -0.1,])
mpp2 = rownames(z[z$PC1 > -0.1 & z$PC2 > 0.09,])
mpp3 = rownames(z[z$PC1 > -0.1 & z$PC2 < 0.09,])

mpppca = z[c(mpp1, mpp2, mpp3),c("PC1", "PC2")]
mpppca$group = c(rep("group1", length(mpp1)), rep("group2", length(mpp2)),
                rep("group3", length(mpp3)))
rownames(mpppca) = gsub("_MPP_MPP", "_MPP", rownames(mpppca))
ggplot(mpppca, aes(x=PC1, y=PC2, color=group)) + geom_point() +
     theme_bw()
```

```{r conversions}
library(biomaRt)
mouse = useMart("ensembl", dataset="mmusculus_gene_ensembl")
conversions = getBM(attributes=c("ensembl_gene_id", "mgi_symbol"), mart=mouse)
```

## differential expression in MPP subpopulations
```{r scde-mpp-setup, cache=FALSE}
cd = counts
cd$id = rownames(cd)
cd = merge(cd, conversions, by.x="id", by.y="ensembl_gene_id")
cd = cd[!duplicated(cd$mgi_symbol),]
rownames(cd) = cd$mgi_symbol
cd$mgi_symbol = NULL
cd$ensmbl_gene_id = NULL
cd$id = NULL
cd = cd[,colnames(cd) %in% rownames(mpppca)]
cd = cd[, colSums(cd > 0)>2000]
cd = cd[rowSums(cd) > 10,]
cd = cd[rowSums(cd > 0) > 5,]
```

```{r scde-mpp, eval=!file.exists("error_models-mpp.RData")}
o.ifm <- scde.error.models(counts = cd, groups = mpppca$group, n.cores = 1,
threshold.segmentation = TRUE, save.crossfit.plots = FALSE,
save.model.plots = FALSE, verbose = 1)
save(cd, o.ifm, file="error_models-mpp.RData")
```

```{r scde-mpp-load, cache=FALSE}
load("error_models-mpp.RData")
valid.cells <- o.ifm$corr.a > 0
o.ifm = o.ifm[valid.cells,]
o.prior <- scde.expression.prior(models = o.ifm, counts = cd, length.out = 400,
 show.plot=TRUE)
groups = mpppca$group
names(groups) = row.names(o.ifm)
```

```{r g1-vs-g2, results='asis', cache=FALSE}
g1_vs_g2 = factor(c(groups[groups == "group1"], groups[groups == "group2"]))
g1_vs_g2_counts = cd[, names(g1_vs_g2)]
g1_vs_g2_ifm = o.ifm[names(g1_vs_g2),]
ediff <- scde.expression.difference(g1_vs_g2_ifm, g1_vs_g2_counts,
o.prior, groups = g1_vs_g2, n.randomizations  =  100, n.cores  =  1, verbose  =  1)
ediff$pvalue = pnorm(-(abs(ediff$Z))) * 2
ediff$padj = p.adjust(ediff$pvalue)
ediff = ediff[order(ediff$pvalue),]
knitr::kable(head(ediff, 25))
```

```{r g1-vs-g3, results='asis', cache=FALSE}
g1_vs_g3 = factor(c(groups[groups == "group1"], groups[groups == "group3"]))
g1_vs_g3_counts = cd[, names(g1_vs_g3)]
g1_vs_g3_ifm = o.ifm[names(g1_vs_g3),]
ediff <- scde.expression.difference(g1_vs_g3_ifm, g1_vs_g3_counts,
o.prior, groups = g1_vs_g3, n.randomizations  =  100, n.cores  =  1, verbose  =  1)
ediff$pvalue = pnorm(-(abs(ediff$Z))) * 2
ediff$padj = p.adjust(ediff$pvalue)
ediff = ediff[order(ediff$pvalue),]
knitr::kable(head(ediff, 25))
```

```{r g2-vs-g3, results='asis'}
g2_vs_g3 = factor(c(groups[groups == "group2"], groups[groups == "group3"]))
g2_vs_g3_counts = cd[, names(g2_vs_g3)]
g2_vs_g3_ifm = o.ifm[names(g2_vs_g3),]
ediff <- scde.expression.difference(g2_vs_g3_ifm, g2_vs_g3_counts,
o.prior, groups = g2_vs_g3, n.randomizations  =  100, n.cores  =  1, verbose  =  1)
ediff$pvalue = pnorm(-(abs(ediff$Z))) * 2
ediff$padj = p.adjust(ediff$pvalue)
ediff = ediff[order(ediff$pvalue),]
knitr::kable(head(ediff, 25))
```

These results aren't too impressive.

## heterogeneity analysis of MPP subpopulations
We could look for sets of genes that show more variation than expected in the
three sets of mpp cells. We do this with Pagoda, in the SCDE package. The idea
behind this is that we might be able to pull out more signal by looking at sets
of genes amongst the three groups. I did this for the MPP populations using
Pagoda, but the results were not very good. We'll use GO terms as the
gene sets here.

```{r pagoda}
cd = scounts
cd = cd[,colnames(cd) %in% rownames(mpppca)]
cd = cd[, colSums(cd > 0)>2000]
cd = cd[rowSums(cd) > 10,]
cd = cd[rowSums(cd > 0) > 5,]
save(cd, mpppca, file="mpp.rdata")
```
```{r pagoda-dont-eval, eval=FALSE}
knn = knn.error.models(cd, groups=mpppca$group, k=ncol(cd)/3, n.cores=2,
                       min.count.threshold=2, min.nonfailed=5, max.model.plots=1)
varinfo <- pagoda.varnorm(knn, counts = cd, trim = 3/ncol(cd), max.adj.var = 5, n.cores = 1, plot = true)
varinfo <- pagoda.subtract.aspect(varinfo, colsums(cd[, rownames(knn)]>0))
```

```{r biomart-go, eval=FALSE}
library(biomart)
library(go.db)
ensembl = usemart("ensembl", dataset = "mmusculus_gene_ensembl")

# constructs a dataframe with two columns: hgnc_symbol and go_id
# if rownames are ensembl ids, use ensembl_gene_id as filter value
go = getbm(attributes = c("mgi_symbol", "go_id"), filters = "mgi_symbol", values = rownames(cd), mart = ensembl)

# use the go.db library to add a column with the go-term to the dataframe
go$term = term(go$go_id)

# create a named list of character vectors out of the df
s = split(go$mgi_symbol, paste(go$go_id,go$term))

# saves the list as a r environment
go.env = list2env(s)
```

Here we look at the magnitude of the first principal component for each gene set
defined in the GO ontologies:

```{r long-use-cached, eval=FALSE}
pwpca <- pagoda.pathway.wPCA(varinfo, go.env, n.components = 1, n.cores = 1, n.internal.shuffles = 0)
clpca <- pagoda.gene.clusters(varinfo, trim = 7.1/ncol(varinfo$mat), n.clusters = 150, n.cores = 1, plot = true)
```

```{r load-precalculated-pagoda}
load("mpp_knn.rdata")
load("mpp_cluster.rdata")
```

```{r pagoda-hc}
df <- pagoda.top.aspects(pwpca, clpca, return.table = TRUE, plot = TRUE, z.score = 1.96)
head(df)
# get full info on the top aspects
tam <- pagoda.top.aspects(pwpca, clpca, n.cells = NULL, z.score = qnorm(0.01/2, lower.tail = FALSE))
hc <- pagoda.cluster.cells(tam, varinfo)
hc2 = hc
hc2$labels = mpppca[hc2$order,]$group
plot(hc2)
```

Breaking it up by pathways doesn't identify these three groups very well. It
looks like there is a block of pathways that might separate out part of
group 1, but those are GO terms that share many of the same genes. The
heatmap below shows the most significant pathways if we take into account
the sharing across the genes. They don't do a great job picking out
the subgroups (red is group 1, blue is group2 and green is group3).

```{r pagoda-aspects}
tamr <- pagoda.reduce.loading.redundancy(tam, pwpca, clpca)
tamr2 <- pagoda.reduce.redundancy(tamr, distance.threshold = 0.9, plot = TRUE, cell.clustering = hc, labrow = NA, labcol = NA, box = TRUE, margins = c(0.5, 0.5), trim = 0,)
col.cols <- rbind(groups = cutree(hc, 3))
l2cols <- c("red", "blue", "green")[factor(mpppca$group, levels=c("group1", "group2", "group3"))]
names(l2cols) = rownames(mpppca)
pagoda.view.aspects(tamr2, cell.clustering = hc, box = TRUE, labcol = NA, margins = c(0.5, 20), col.cols=l2cols)
```

## wrapup
It seems like for single cell analyses we need a large number of cells if the
goal is to do any kind of heterogeneity analysis within a subpopulation. Most
of the successful single-cell papers are solving a much easier problem, looking
at different cell types which are much easier to separate.

[This](http://genome.cshlp.org/content/early/2015/10/01/gr.192237.115.abstract)
paper that Agnieska sent has a similar type of experiment, also using SMART-seq data.
They sequenced about 1,000 cells and were looking at LT-HSCS, ST-HSCS and MPPs.
I think this scale of doing single-cell sequencing would be more appropriate
to do the analyses of unknown subpopulations.
